### PR TITLE
Deprecate `InitArguments` struct and replace it with `InitializationSettings`

### PR DESCRIPTION
**Pull request by dalg24**
I drafted a replacement struct for `InitArguments`.  ~~At this time it is still missing the implicit conversion from `InitArguments`.~~  I haven't put any thought into the name yet.  I am looking for early feedback on the design, that is public member function to set the values that can be chained and "private" macros to query if a variable is set and and get its value.

### CODE REVIEW

File: core/unit_test/default/TestDefaultDeviceDevelop.cpp [hunk 1]
```diff
@@ -49,8 +49,26 @@
...
 
-TEST(defaultdevicetype, development_test) {}
+namespace {
 
-}  // namespace Test
+TEST(foo, initialization_arguments) {
+  auto ia = Kokkos::InitializationArguments()
+                .set_num_threads(255)
+                .set_disable_warnings(true)
+                .set_tool_lib("my_custom_tool.so");
```
[the remark this answers is not in the document]
Comment (cz4rs) [reply core/unit_test/default/TestDefaultDeviceDevelop.cpp:59]: Nice to have category: for each of these methods we could have a docstring matching what we have in Kokkos command line help message (for that option).

**Review by masterleinad — COMMENTED** [review]
I think this is conceptionally (interface-wise) good. I'm not quite sure if we really need the getter-functions but they seem to be good for testing at least.
As discussed elsewhere, we also need to see what issues we would need to revolve for interfaces taking the `InitArguments` right now.

**Comment by cz4rs** [comment]
I would be in favor of adding the getters as well. This obviously gets complicated for the cases where the variable is not set (what to return?).

**Comment by nliber** [comment]
I like this interface as well.  

While I don't see a need for getters, maybe a way to print out the configuration?

**Comment by PhilMiller** [comment]
I would favor getters and making the members private, so that the compiler will enforce that all modifications go through the setters, rather than doing weird stuff with clunky name modifications in macros.

**Review by masterleinad — COMMENTED** [review]
Looks like a guard for the deprecated struct is missing:
```
/var/jenkins/workspace/Kokkos/core/unit_test/TestInitializationSettings.cpp:56:11: error: no type named 'InitArguments' in namespace 'Kokkos'
  Kokkos::InitArguments arguments;
  ~~~~~~~~^
1 error generated.
core/unit_test/CMakeFiles/KokkosCore_UnitTest_Default.dir/build.make:94: recipe for target 'core/unit_test/CMakeFiles/KokkosCore_UnitTest_Default.dir/TestInitializationSettings.cpp.o' failed
```

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.cpp [hunk 2]
```diff
@@ -167,11 +167,12 @@ int g_openmptarget_space_factory_initialized =
     Kokkos::Impl::initialize_space_factory<OpenMPTargetSpaceInitializer>(
         "160_OpenMPTarget");
 
-void OpenMPTargetSpaceInitializer::initialize(const InitArguments& args) {
-  // Prevent "unused variable" warning for 'args' input struct.  If
+void OpenMPTargetSpaceInitializer::initialize(
+    const InitializationSettings& settings) {
+  // Prevent "unused variable" warning for 'settings' input struct.  If
   // Serial::initialize() ever needs to take arguments from the input
```
Comment (Rombur) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.cpp:173]: Wrong comment. It should be `OpenMPTarget::initialize()`

File: core/src/impl/Kokkos_Core.cpp [hunk 3]
```diff
@@ -100,45 +99,105 @@ void declare_configuration_metadata(const std::string& category,
   metadata_map[category][key] = value;
 }
 
-ExecSpaceManager& ExecSpaceManager::get_instance() {
+void combine(Kokkos::InitializationSettings& out,
+             Kokkos::InitializationSettings const& in) {
```
Comment (Rombur) [inline core/src/impl/Kokkos_Core.cpp:103]: Why do you need this function?

**Review by Rombur — APPROVED** [review]

File: core/src/impl/Kokkos_InitializationSettings.hpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+#undef KOKKOS_IMPL_INIT_ARGS_DATA_MEMBER
+#undef KOKKOS_IMPL_DECLARE
+
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
+ public:
+  InitializationSettings() = default;
+
+  InitializationSettings(InitArguments const& old) {
+    if (old.num_threads != -1) {
+      (void)set_num_threads(old.num_threads);
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_InitializationSettings.hpp:158]: I don't think so

File: core/src/impl/Kokkos_Core.cpp [hunk 5]
```diff
@@ -1046,3 +1054,7 @@ void _kokkos_pgi_compiler_bug_workaround() {}
 #endif
 
 }  // namespace Kokkos
+
+Kokkos::Impl::InitializationSettingsHelper<std::string>::storage_type const
+    Kokkos::Impl::InitializationSettingsHelper<std::string>::unspecified =
+        "some string we don't expect user would ever provide";
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_Core.cpp:1060]: Considering that we will drop it when C++17 is available, I'd rather not create a source file.

File: core/src/Kokkos_Core_fwd.hpp [hunk 6]
```diff
@@ -94,7 +94,10 @@ template <class ExecutionSpace, class MemorySpace>
 struct Device;
 
 // forward declare here so that backend initializer calls can use it.
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
 struct InitArguments;
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_Core_fwd.hpp:98]: No that would raise warning in the converting constructor `InitializationSettings(InitArguments)`

File: core/src/impl/Kokkos_Core.cpp (hunk 3 above)
Comment (dalg24) [reply core/src/impl/Kokkos_Core.cpp:103]: It replaces `void combine(InitArguments& out, InitArguments const& in)` and we use it in `initialize(InitializationSettings const&)`

File: core/src/impl/Kokkos_Core.cpp [hunk 7]
```diff
@@ -100,45 +99,105 @@ void declare_configuration_metadata(const std::string& category,
...
+  KOKKOS_IMPL_COMBINE_SETTING(num_threads);
+  KOKKOS_IMPL_COMBINE_SETTING(device_id);
+  KOKKOS_IMPL_COMBINE_SETTING(num_devices);
+  KOKKOS_IMPL_COMBINE_SETTING(skip_device);
+  KOKKOS_IMPL_COMBINE_SETTING(disable_warnings);
+  KOKKOS_IMPL_COMBINE_SETTING(tune_internals);
+  KOKKOS_IMPL_COMBINE_SETTING(tool_help);
+  KOKKOS_IMPL_COMBINE_SETTING(tool_lib);
+  KOKKOS_IMPL_COMBINE_SETTING(tool_args);
+#undef KOKKOS_IMPL_COMBINE_SETTING
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_Core.cpp:118]: Not changing as I do not agree with the suggestion

File: core/src/impl/Kokkos_InitializationSettings.hpp [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+  static storage_type const unspecified;
+};
+}  // namespace Impl
+
+class InitializationSettings {
+#define KOKKOS_IMPL_INIT_ARGS_DATA_MEMBER(NAME) \
+  impl_do_not_use_i_really_mean_it_##NAME##_
+
+#define KOKKOS_IMPL_INIT_ARGS_DATA_MEMBER_TYPE(NAME) impl_##NAME##_type
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_InitializationSettings.hpp:112]: This was introduced before we added the getters and was really handy then.  I would rather wait until C++17 is available to refactor.  Then I will get rid of the member type and "inline" all data member references.

File: core/src/Serial/Kokkos_Serial.cpp [hunk 9]
```diff
@@ -194,11 +194,12 @@ namespace Impl {
...
 
-void SerialSpaceInitializer::initialize(const InitArguments& args) {
-  // Prevent "unused variable" warning for 'args' input struct.  If
+void SerialSpaceInitializer::initialize(
+    const InitializationSettings& settings) {
+  // Prevent "unused variable" warning for 'settings' input struct.  If
   // Serial::initialize() ever needs to take arguments from the input
   // struct, you may remove this line of code.
-  (void)args;
+  (void)settings;
```
Comment (nliber) [inline core/src/Serial/Kokkos_Serial.cpp:202]: Minor: I know it wasn't done like this originally, but why not just comment out the variable name in the function declaration?

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
OpenMPTarget timeout is clearly unrelated

**Pull request merged**